### PR TITLE
Feature-1645-heaptimer

### PR DESCRIPTION
**Pull request by KaoImin**
add  a min-heap timer

### CODE REVIEW

File: src/core/votetime.rs [hunk 1]
```diff
@@ -35,39 +37,46 @@ pub struct TimeoutInfo {
...
         loop {
-            select! {
-                settime = innersetter.recv() =>  {
-                    let oksettime = settime.unwrap();
-                    let notify = self.timer_notify.clone();
+            // take the peek of the min-heap-timer sub now as the sleep time otherwise set timeout as 0
+            let timeout = if !timer_heap.is_empty() {
+                timer_heap.peek_min().cloned().unwrap() - Instant::now()
+            } else {
+                Duration::from_millis(0)
```
Comment (jerry-yu) [inline src/core/votetime.rs]: here should set a long duraton eg. 100s  or sync wait

File: src/core/votetime.rs [hunk 2]
```diff
@@ -35,39 +37,46 @@ pub struct TimeoutInfo {
...
-                            trace!(" ************ {:?}",oksettime);
-                            thread::sleep(oksettime.timeval);
-                            notify.send(oksettime).unwrap();
-                        });
-                    }
+            if !timer_heap.is_empty() {
+                let now = Instant::now();
+                let notify = self.timer_notify.clone();
+                // if some timers are set as the same time, send timeout messages and pop them
+                while now >= timer_heap.peek_min().cloned().unwrap() {
```
Comment (jerry-yu) [inline src/core/votetime.rs]: unwrap 可能导致panic吧？

File: src/core/votetime.rs (hunk 1 above)
Comment (KaoImin) [reply src/core/votetime.rs]: yes

File: src/core/votetime.rs [hunk 3]
```diff
@@ -17,15 +17,17 @@
...
 use std::sync::mpsc::{Receiver, Sender};
-use std::thread;
-use std::time::Duration;
-use threadpool::ThreadPool;
+use std::time::{Duration, Instant};
+// use std::thread
+// use threadpool::ThreadPool;
 
-const THREAD_POOL_NUM: usize = 10;
+extern crate min_max_heap;
```
Comment (jerry-yu) [inline src/core/votetime.rs]: 这个 extern放在最上面吧

Comment (KaoImin) [reply src/core/votetime.rs]: ok

File: src/core/votetime.rs (hunk 2 above)
Comment (jerry-yu) [reply src/core/votetime.rs]: while let Some( elem) = timer_heap.peek_min( {
let now = Instant::now();
let notify = self.timer_notify.clone();
   if now > = elem {
send
pop
}
}

Comment (KaoImin) [reply src/core/votetime.rs]: This method breaks the rust borrow rule, and has to use other API instead.

**Review by driftluo — APPROVED** [review]

File: src/core/votetime.rs [hunk 4]
```diff
@@ -35,39 +33,51 @@ pub struct TimeoutInfo {
...
+        let mut timeout_info = HashMap::new();
+
         loop {
-            select! {
-                settime = innersetter.recv() =>  {
-                    let oksettime = settime.unwrap();
-                    let notify = self.timer_notify.clone();
+            // take the peek of the min-heap-timer sub now as the sleep time otherwise set timeout as 100
+            let timeout = if !timer_heap.is_empty() {
+                *timer_heap.peek_min().unwrap() - Instant::now()
```
Comment (jerry-yu) [inline src/core/votetime.rs:58]: 会检查出负的结果！？

File: src/core/votetime.rs [hunk 5]
```diff
@@ -35,39 +33,51 @@ pub struct TimeoutInfo {
...
             timer_seter: rs,
-            thpool: pool,
+            // thpool: pool,
         }
     }
 
     pub fn start(&self) {
-        let innersetter = &self.timer_seter;
-        let zero_time = ::std::time::Duration::new(0, 0);
+        let mut timer_heap = MinMaxHeap::new();
```
Comment (jerry-yu) [inline src/core/votetime.rs:52]: 为啥不直接把timeoutInfo放进去，现在只是放过time val

Comment (KaoImin) [reply src/core/votetime.rs:52]: timeoutinfo 放进最小堆没法排序，就另开了一个 HashMap 通过 timeval 查找 timeoutinfo

Comment (driftluo) [reply src/core/votetime.rs:52]: 其实可以做到排序的，自己对 timeoutinfo 实现 Eq，Ord

Comment (KaoImin) [reply src/core/votetime.rs:52]: 对，那样需要重载运算符

**Review by u2 — APPROVED** [review]

**Pull request merged**